### PR TITLE
Adding output section 

### PR DESCRIPTION
**Pull request by alexcos20**
Adding output section to be passed to brizo

Similar:
 [ ] https://github.com/oceanprotocol/squid-py/pull/431

Blocking:
 [ ] https://github.com/oceanprotocol/brizo/pull/182

### CODE REVIEW

File: src/brizo/Brizo.ts [hunk 1]
```diff
@@ -116,7 +116,8 @@ export class Brizo extends Instantiable {
         consumerAccount: Account,
         algorithmDid?: string,
         algorithmMeta?: MetaData,
-        jobId?: string
+        jobId?: string,
+        output?: Object
```
Comment (kremalicious) [inline src/brizo/Brizo.ts]: `Object` should be actually defined as an interface and then referenced here and everywhere else it's being used, where did fa09f6f2c76ee3f193b10eeb6e3280c58501d6a9 go?

File: src/ocean/OceanCompute.ts [hunk 2]
```diff
@@ -74,26 +74,72 @@ export class OceanCompute extends Instantiable {
         })
     }
 
+    /**
+     * return minimal output object
```
Comment (kremalicious) [inline src/ocean/OceanCompute.ts]: squid-js should not be responsible validating configs and to set defaults for the whole object, Brizo should do that

File: src/ocean/OceanCompute.ts [hunk 3]
```diff
@@ -74,26 +74,72 @@ export class OceanCompute extends Instantiable {
...
+        if (!output){
+            //build a minimal object and return it
+            return (await this.minimal_output_object())
+        }
+        if((!output["publishAlgorithmLog"] || output["publishAlgorithmLog"]==false) && (!output["publishOutput"] || output["publishOutput"]==false)){
+            return (await this.minimal_output_object())
+        }
+        if(!output["brizoAddress"])
+            output["brizoAddress"]=this.config.brizoAddress
+        if(!output["brizoUrl"])
```
Comment (kremalicious) [inline src/ocean/OceanCompute.ts]: regardless, these values do not follow what's being worked on in https://github.com/oceanprotocol/dev-ocean/pull/159, see https://github.com/oceanprotocol/dev-ocean/blob/feature/squid-compute/doc/architecture/squid.md#start (e.g. we use `Uri` for all values)

I strongly suggest first agreeing on and merging specs before implementing them, so we should first get that spec over there merged and then we can implement

File: src/brizo/Brizo.ts (hunk 1 above)
Comment (alexcos20) [reply src/brizo/Brizo.ts]: fixed

File: src/ocean/OceanCompute.ts (hunk 2 above)
Comment (alexcos20) [reply src/ocean/OceanCompute.ts]: The Output section contains aquariusUri, which is not available to Brizo.  So it must be done one level above.  After all, it's user's choice where and when to publish the result.

He must be able to do compute on marketplace A (aqua_A, brizo_A, even network_A) and then to publish the results to marketplace_B(aqua_B, brizo_B, even network_B)

Not to mention metadata & whitelist attributes, which are strictly user defined.

File: src/ocean/OceanCompute.ts (hunk 3 above)
Comment (alexcos20) [reply src/ocean/OceanCompute.ts]: fixed

File: src/ocean/OceanCompute.ts [hunk 4]
```diff
@@ -74,26 +87,66 @@ export class OceanCompute extends Instantiable {
...
+        return minimaloutput as Output
+    }
+
+    /**
+     * Check the output object and add default properties if needed
+     * @param  {Account} consumerAccount The account of the consumer ordering the service.
+     * @param  {Object} output Output section used for publishing the result.
+     * @return {Promise<Output>} Returns output object
+     */
+    public async checkoutput(consumerAccount: Account, output?: Output): Promise<Output> {
```
Comment (kremalicious) [inline src/ocean/OceanCompute.ts]: If we want the default setting happening on squid side, then I think handling the whole check in one method, made more readable by following object structure so it's more clear what's happening in the object changing would be nicer. Also no need for async promise function (I think, not sure, but restructuring an object is not something we have to wait for), all together something like:

```js
private checkOutput(consumerAccount: Account, output?: Output): Output {
    const isDefault =
        !output || (!output.publishAlgorithmLog && !output.publishOutput)

    if (isDefault) {
        return {
            publishAlgorithmLog: false,
            publishOutput: false
        }
    }

    return {
        publishAlgorithmLog: output.publishAlgorithmLog,
        publishOutput: output.publishOutput,
        brizoAddress: output.brizoAddress || this.config.brizoAddress,
        brizoUri: output.brizoUri || this.config.brizoUri,
        metadataUri: output.metadataUri || this.config.aquariusUri,
        nodeUri: output.nodeUri || this.config.nodeUri,
        owner: output.owner || consumerAccount.getId(),
        secretStoreUri: output.secretStoreUri || this.config.secretStoreUri
    }
}
```

File: src/ocean/OceanCompute.ts [hunk 5]
```diff
@@ -17,6 +17,19 @@ export enum ComputeJobStatus {
     Deleted
 }
 
+export interface Output {
+    publishAlgorithmLog?: boolean
+    publishOutput?: boolean
+    brizoAddress?: string
+    brizoUri?: string
+    metadata?: Record<string, any>
```
Comment (kremalicious) [inline src/ocean/OceanCompute.ts]:
```suggestion
    metadata?: MetaData
```

File: src/ocean/OceanCompute.ts [hunk 6]
```diff
@@ -17,6 +17,19 @@ export enum ComputeJobStatus {
...
+    metadataUri?: string
+    nodeUri?: string
+    owner?: string
+    secretStoreUri?: string
+    whitelist?: Array<string>
```
Comment (kremalicious) [inline src/ocean/OceanCompute.ts]: just for consistency, we always do this for array of strings:

```suggestion
    whitelist?: string[]
```

File: src/ocean/OceanCompute.ts (hunk 4 above)
Comment (alexcos20) [reply src/ocean/OceanCompute.ts]: thanks

File: src/ocean/OceanCompute.ts [hunk 7]
```diff
@@ -74,26 +87,72 @@ export class OceanCompute extends Instantiable {
         })
     }
 
+    /**
+     * return minimal output object
+     * @return {Promise<Object>} Returns output object
+     */
+    public async getminimaloutput(): Promise<Output> {
```
Comment (kremalicious) [inline src/ocean/OceanCompute.ts]: where is this method still in use?

File: src/ocean/OceanCompute.ts [hunk 8]
```diff
@@ -74,26 +87,72 @@ export class OceanCompute extends Instantiable {
...
+            publishAlgorithmLog: false,
+            publishOutput: false
+        }
+        return minimaloutput as Output
+    }
+
+    /**
+     * Check the output object and add default properties if needed
+     * @param  {Account} consumerAccount The account of the consumer ordering the service.
+     * @param  {Object} output Output section used for publishing the result.
```
Comment (kremalicious) [inline src/ocean/OceanCompute.ts]:
```suggestion
     * @param  {Output} output Output section used for publishing the result.
```

File: src/ocean/OceanCompute.ts [hunk 9]
```diff
@@ -74,26 +87,72 @@ export class OceanCompute extends Instantiable {
...
+        }
+    }
+
     /**
      * Start the execution of a compute job.
      * @param  {Account} consumerAccount The account of the consumer ordering the service.
      * @param  {string} agreementId The service agreement ID.
      * @param  {string} algorithmDid The DID of the algorithm asset (of type `algorithm`) to run on the asset.
      * @param  {MetaData} algorithmMeta Metadata about the algorithm being run if `algorithm` is being used. This is ignored when `algorithmDid` is specified.
+     * @param  {Object} output Output section used for publishing the result.
```
Comment (kremalicious) [inline src/ocean/OceanCompute.ts]:
```suggestion
     * @param  {Output} output Define algorithm output publishing. Publishing the result of a compute job is turned off by default.
```

File: src/ocean/OceanCompute.ts (hunk 7 above)
Comment (alexcos20) [reply src/ocean/OceanCompute.ts]: removed

**Pull request merged**